src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java

```diff
@@ -83,44 +89,62 @@ public void close() {
+    public void publish(ClusterChangedEvent clusterChangedEvent, final Discovery.AckListener ackListener) {
+        Set<DiscoveryNode> nodesToPublishTo = new HashSet<>(clusterChangedEvent.state().nodes().size());
         DiscoveryNode localNode = nodesProvider.nodes().localNode();
-        for (final DiscoveryNode node : clusterState.nodes()) {
+        for (final DiscoveryNode node : clusterChangedEvent.state().nodes()) {
             if (node.equals(localNode)) {
                 continue;
             }
             nodesToPublishTo.add(node);
         }
-        publish(clusterState, nodesToPublishTo, new AckClusterStatePublishResponseHandler(nodesToPublishTo, ackListener));
+        publish(clusterChangedEvent, nodesToPublishTo, new AckClusterStatePublishResponseHandler(nodesToPublishTo, ackListener));
     }
 
-    private void publish(final ClusterState clusterState, final Set<DiscoveryNode> nodesToPublishTo,
+    private void publish(final ClusterChangedEvent clusterChangedEvent, final Set<DiscoveryNode> nodesToPublishTo,
                          final BlockingClusterStatePublishResponseHandler publishResponseHandler) {
 
         Map<Version, BytesReference> serializedStates = Maps.newHashMap();
+        Map<Version, BytesReference> serializedDiffs = Maps.newHashMap();
 
         final AtomicBoolean timedOutWaitingForNodes = new AtomicBoolean(false);
         final TimeValue publishTimeout = discoverySettings.getPublishTimeout();
+        final boolean sendFullVersion = !discoverySettings.getPublishDiffs() || clusterChangedEvent.previousState() == null;
+        final ClusterState clusterState = clusterChangedEvent.state();
+        Diff diff = null;
 
         for (final DiscoveryNode node : nodesToPublishTo) {
 
             // try and serialize the cluster state once (or per version), so we don't serialize it
             // per node when we send it over the wire, compress it while we are at it...
-            BytesReference bytes = serializedStates.get(node.version());
-            if (bytes == null) {
-                try {
-                    BytesStreamOutput bStream = new BytesStreamOutput();
-                    StreamOutput stream = CompressorFactory.defaultCompressor().streamOutput(bStream);
-                    stream.setVersion(node.version());
-                    ClusterState.Builder.writeTo(clusterState, stream);
-                    stream.close();
-                    bytes = bStream.bytes();
-                    serializedStates.put(node.version(), bytes);
-                } catch (Throwable e) {
-                    logger.warn("failed to serialize cluster_state before publishing it to node {}", e, node);
-                    publishResponseHandler.onFailure(node, e);
-                    continue;
+            BytesReference bytes;
+            boolean newlyAddedNode = !clusterChangedEvent.previousState().nodes().nodeExists(node.id());
+            if (sendFullVersion || newlyAddedNode) {
```

Comment:
I think this code will be simpler if we have two methods - sendFullClusterState and sendClusterStateDiff , each dealing with it's own serialization (and have the cache map passed to them). We can have sendClusterStateDiff fall back to sendFullClusterState if needed , which will mean no resend method..